core/corehttp/gateway_handler.go

```diff
@@ -154,27 +157,19 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 		ipnsHostname = true
 	}
 
-	p, err := path.ParsePath(urlPath)
-	if err != nil {
-		webError(w, "Invalid Path Error", err, http.StatusBadRequest)
-		return
-	}
-
-	nd, err := core.Resolve(ctx, i.node.Namesys, i.node.Resolver, p)
-	// If node is in offline mode the error code and message should be different
-	if err == core.ErrNoNamesys && !i.node.OnlineMode() {
+	dr, err := i.api.Cat(ctx, urlPath)
+	dir := false
+	if err == coreiface.ErrOffline {
 		w.WriteHeader(http.StatusServiceUnavailable)
 		fmt.Fprint(w, "Could not resolve path. Node is in offline mode.")
 		return
+	} else if err == coreiface.ErrIsDir {
+		dir = true
 	} else if err != nil {
 		webError(w, "Path Resolve error", err, http.StatusBadRequest)
 		return
-	}
-
-	pbnd, ok := nd.(*dag.ProtoNode)
-	if !ok {
-		webError(w, "Cannot read non protobuf nodes through gateway", dag.ErrNotProtobuf, http.StatusBadRequest)
-		return
+	} else {
```

Comment:
One of the conditionals above doesn't return (`else if err == coreiface.ErrIsDir`), so we'd be calling `nil.Close()`. This piece first tries a reader, then lists links if it's a directory.